Potential EX gym detection

KartulUdus opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
The generally accepted theory on where EX raids appear is OSM data light green parks. 
This PR gives the ability to store which gyms can possibly spawn an EX raid within an area
## Description
<!--- Describe your changes in detail -->

* New column in gym table (boolean for park or not)
* New requirement for overpy (pip install -r requirements.txt --upgrade)
* New arg for populating Gym.park

Uses existing geofencing functions
Using a single area geofence is required 

Launch with 
`python runserver.py -exg -gf geofences/yourgeofence.txt`
Parks and other applicable green areas will be pulled within the bounding box of your geofence.
Existing gyms within the geofence will be compared against areas
Will update Gym.park=True if the gym is inside an area

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
My discord was making CSV files of gyms manually and using tools like overpass-turbo to generate maps with the data.
The query was easy enough to implement with overpy to map directly
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Hardly at all, I mostly photoshopped the examples below.
## Screenshots (if appropriate):
![screen shot 2018-01-16 at 03 09 25](https://user-images.githubusercontent.com/23346150/34967359-b8a2ebec-fa6a-11e7-91c2-25e2ecb7985d.png)

The switch:
![screen shot 2018-01-19 at 00 08 00](https://user-images.githubusercontent.com/23346150/35125058-ec40c2c4-fcb0-11e7-9a7c-df35e71ffa52.png)

Switch when turned off or on:
![screen shot 2018-01-19 at 00 27 58](https://user-images.githubusercontent.com/23346150/35125081-fb626a5a-fcb0-11e7-8874-a6483279f6bd.png)
![screen shot 2018-01-18 at 23 59 35](https://user-images.githubusercontent.com/23346150/35125090-00e3960c-fcb1-11e7-8484-96c30edc6c32.png)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--  NOTE: In order to check code style locally and avoid having your build rejected by Travis, -->
<!--  run the following commands before you commit: `flake8 .` and `npm run lint`. Fix any -->
<!--  issues they point out. Note also that flake's NOQA is disabled on Travis. -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have made the thing work with the front-end

blakewenloe commented:
```
2018-01-15 20:34:32,979 [        MainThread][           osm][    INFO] Finding border points from geofence
2018-01-15 20:34:32,979 [        MainThread][           osm][    INFO] Finding parks within zone
2018-01-15 20:34:51,293 [        MainThread][     runserver][   ERROR] Uncaught exception
Traceback (most recent call last):
  File "runserver.py", line 531, in <module>
    main()
  File "runserver.py", line 268, in main
    exgyms(args.geofence_file)
  File "/home/RocketMap/pogom/osm.py", line 62, in exgyms
    gyms = Gym.get_gyms(south, west, north, east)
  File "/home/RocketMap/pogom/models.py", line 499, in get_gyms
    for g in results:
  File "/usr/local/lib/python2.7/dist-packages/peewee.py", line 2967, in __iter__
    return iter(self.execute())
  File "/usr/local/lib/python2.7/dist-packages/peewee.py", line 2960, in execute
    self._qr = ResultWrapper(model_class, self._execute(), query_meta)
  File "/usr/local/lib/python2.7/dist-packages/peewee.py", line 2656, in _execute
    return self.database.execute_sql(sql, params, self.require_commit)
  File "/usr/local/lib/python2.7/dist-packages/playhouse/shortcuts.py", line 208, in execute_sql
    cursor.execute(sql, params or ())
  File "/usr/lib/python2.7/dist-packages/MySQLdb/cursors.py", line 226, in execute
    self.errorhandler(self, exc, value)
  File "/usr/lib/python2.7/dist-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
OperationalError: (1054, "Unknown column 't1.park' in 'field list'")
```

KartulUdus commented:
there's a database migration, run normally first

blakewenloe commented:
Whoops okay I upgraded. Now i'm getting this error:
```
2018-01-15 20:38:33,709 [        MainThread][           osm][    INFO] Finding border points from geofence
2018-01-15 20:38:33,709 [        MainThread][           osm][    INFO] Finding parks within zone
2018-01-15 20:38:45,531 [        MainThread][     runserver][   ERROR] Uncaught exception
Traceback (most recent call last):
  File "runserver.py", line 531, in <module>
    main()
  File "runserver.py", line 268, in main
    exgyms(args.geofence_file)
  File "/home/RocketMap/pogom/osm.py", line 62, in exgyms
    gyms = Gym.get_gyms(south, west, north, east)
  File "/home/RocketMap/pogom/models.py", line 528, in get_gyms
    p['pokemon_name'] = get_pokemon_name(p['pokemon_id'])
  File "/home/RocketMap/pogom/utils.py", line 872, in get_pokemon_name
    return i8ln(get_pokemon_data(pokemon_id)['name'])
  File "/home/RocketMap/pogom/utils.py", line 862, in get_pokemon_data
    args.root_path,
AttributeError: 'Namespace' object has no attribute 'root_path'
```

blakewenloe commented:
I'm noticing that gyms are being set back to `park=0` after they've been set as `park=1` in the DB. I ran `select name, park from gym join gymdetails on gym.gym_id=gymdetails.gym_id where park=1;` and got 40 in the first set, and then 15 the next time I ran it.

blakewenloe commented on pogom/osm.py:
```diff
@@ -0,0 +1,79 @@
+            # Check if gyms falls within a designated park and update if park
+            if Geofences.is_point_in_polygon_custom(gympoint, data):
+                gymname = Gym.get_gym(gym[0])['name'].encode('utf8')
+                log.info('gym:{} could spawn legendary raid'.format(gymname))
```
Just a nitpick, but they can all technically spawn legendary raids. I'd change this to `could spawn ex raid` or `eligible for ex raid`.

KartulUdus commented:
Interesting, then it seems that it's getting overridden with default once scanned. 
I'll take a look later today

blakewenloe commented:
I may be completely off-center, but do you think that, after running your command, the next time those gyms are scanned (which is pretty quick), they update the park back to default/0?

blakewenloe commented:
I may be completely off-center, but do you think that, after running your command, the next time those gyms are scanned (which is pretty quick), they update the park back to default/0?

blakewenloe commented:
That's something I think I can test. I'll turn off all of my other instances and try it.

fosJoddie commented:
Interesting PR, I'll give this one a test when I can.

If you are scanning normally while this script is running, I'm pretty sure that the changes WILL be overwritten by the normal scanner upserting gyms. Not really a bug if that is the case. This tool should be ran while normal scanner is not running.

jorgitoh commented:
osm.py", line 77, in exgyms
    gymname = Gym.get_gym(gym[0])['name'].encode('utf8')
AttributeError: 'NoneType' object has no attribute 'encode'

I think not all gyms has "name". Maybe we need to run first with -gi to get all the data from gyms?.

blakewenloe commented:
Confirming it works perfectly when I have the other instances (aside from `-os`) turned off.

SkOODaT commented:
its on a gym scan its setting it back to zero its defaulting the value back, im already starting to think how to stop this ..... the check is too much to do entirely every time..... prolly best thing to do is have it to the check every time it scans a gym for detailed infos singularly

jorgitoh commented:
@SkOODaT Same problem. Any instance reset the value of park to 0. Any idea?.

blakewenloe commented:
If workers don't actually pull the park info for the gyms outside of running this particular script, wouldn't it make sense to stop them from updating the parks column when running a normal instance?

SkOODaT commented:
yes but with the way data is gather i dont think its possible to "stop" the update when the worker searches ... my gym battleing stuff has this same issue, there for populating the gym table with a singular check uesing overpy to osm would make sense it would just update the table with the data everytime, i guess it might be a slight waste of resource tho, im no pro coder

jorgitoh commented:
Im comparing the result with http://overpass-turbo.eu/ and current invitation for ex raids. For example. This code return false on http://www.openstreetmap.org/way/9939231

KartulUdus commented:
@SkOODaT rerunning it every time would be a quick way to get banned from OSM I believe.
@jorgitoh My sources claim that the EX raids are currently based on green parks from 2016 july, could you add that to your query to see if the park is there?

jorgitoh commented:
@KartulUdus its weird. Im using this
http://overpass-turbo.eu/

The query its similar to ur query:

[out:json][timeout:25];
// gather results
(
  // query part for: “leisure=park”
  node["leisure"="park"]({{bbox}});
  way["leisure"="park"]({{bbox}});
  relation["leisure"="park"]({{bbox}});
  // query part for: “leisure=playground”
  node["leisure"="playground"]({{bbox}});
  way["leisure"="playground"]({{bbox}});
  relation["leisure"="playground"]({{bbox}});
  // query part for: “leisure=recreation_ground”
  node["leisure"="recreation_ground"]({{bbox}});
  way["leisure"="recreation_ground"]({{bbox}});
  relation["leisure"="recreation_ground"]({{bbox}});
  // query part for: “landuse=recreation_ground”
  node["landuse"="recreation_ground"]({{bbox}});
  way["landuse"="recreation_ground"]({{bbox}});
  relation["landuse"="recreation_ground"]({{bbox}});
  // query part for: “landuse=grass”
  node["landuse"="grass"]({{bbox}});
  way["landuse"="grass"]({{bbox}});
  relation["landuse"="grass"]({{bbox}});
  // query part for: “landuse=meadow”
  node["landuse"="meadow"]({{bbox}});
  way["landuse"="meadow"]({{bbox}});
  relation["landuse"="meadow"]({{bbox}});
  // query part for: “landuse=greenfield”
  node["landuse"="greenfield"]({{bbox}});
  way["landuse"="greenfield"]({{bbox}});
  relation["landuse"="greenfield"]({{bbox}});
);
// print results
out body;
>;
out skel qt;

I have a few gyms in park with values in false

SkOODaT commented:
I was just uesing the one as example the rest need to be added of course lol ... And are you shure it cause a ban ...if so then we need say a json to store the data in and call on it I'm pretty sure your not going to be able to stop the gym table from zeroing the data without data to replace it :)

KartulUdus commented:
@jorgitoh could you add `[date:"2016-07-10T00:00:00Z"]` to the top of your query and see if it was a park then?

I got an idea how to solve the returning to 0 but only an idea so far ;)

jorgitoh commented:
Yes, with ur the date the website http://overpass-turbo.eu/ shows the park correctly:

![worksfine](https://user-images.githubusercontent.com/87159/34983296-cd94449e-fa8b-11e7-9f21-c73bf7bbd972.png)

Maybe an error in the code?

friscoMad commented on pogom/osm.py:
```diff
@@ -0,0 +1,79 @@
+
+args = get_args
+
+app = Flask(__name__)
```
app can be None in this case not need to instance Flask

KartulUdus replied:
You are indeed correct sir, changed now

friscoMad commented:
Due to the current replace-into query in bulk updater there is no way to keep the park field in the db, hopefully the db insert change comming after devkat exclusive time can keep that info, otherwise we would need to change how we insert the gyms in db or do a query before any gym insert.

On the other hand as I said in a previous comment, I prefer to not add more params for tools that are only done once a external script seems a better fit. On the other hand we can make this to run every day or so to keep data updated and avoid the param.

pogo-excalibur commented:
Would it not possibly have been easier to have done this as a front end change, using an overlay layer to indicate nearby parks (with the determination of EX-elegible gyms left to users)?

What advantage are you aiming for by having this done at the backend/db level?

KartulUdus commented:
* changed log message to `<Gym> is eligible for legendary raid`
 * removed Flask from osm.py
 * changed log message so that by default it would use gym_name, but if none is found (args.gym_info == false) it defaults back to gym_id
* up to date with current develop  
* applied forehead to hard surface for how to pass old Gym.parks value along the insert, alas to no avail

SkOODaT commented:
Here you all go i slaved away on some Sql :S lol
The Fix add this right above your gym webhooks in models.py

```
                with Gym.database().execution_context():
                    Query = Gym.select().where(Gym.gym_id == f.id).dicts()
                    park_id = None
                    for gym in list(Query):
                        park_id = gym['park']
                    log.debug(park_id)
```
and add 
```
                        'park':
                            park_id
```
onto the webhooks and database stuff
it will preserve whatever is on the gym table

SkOODaT commented:
oh also remove 
park = BooleanField()
you dont need it

KartulUdus commented:
thats what I was _trying_ to do @SkOODaT 😄 , query it before making new insert and use old value. 

It is an extra query which then hopefully can be removed after the 26/01/2018 magic.
Still need a switch for the FE to only show these gyms, stay tuned

SkOODaT commented:
i have a feature suggestion too, now what we need is a way for it to just check on server startup every 30 days or something that shouldn't be too hard... some front end work would be nice as well some way for gyms to be distinguished maby we need a potiential ex raid gym icon or something 
throwing some ideas out there :)

SkOODaT commented:
`'park': park_id `needs to be added onto eggs/raids too i missed that ..... and if you want it in the extended gym info webhooks it needs to be but again down in the gym phaseing

SkOODaT commented:
You should also take A.R.T and I's Weather S2Cells and mod it to work with raid S2cells show then on the front end when your new switch is on

KartulUdus commented:
will add to eggs/raid possibly today 

I think we shouldn't add too many switches tho. perhaps cell size could be a dropdown in upcoming weather pr #2420 ?

blakewenloe commented:
I agree. Adding cell size here is probably cramming too many features into one PR.

SkOODaT commented:
well the only reason i say put it into OSM is because A.R.T has the weather stuff all separated into weather.js and weather.py .... also the reason adding the ex raid s2cells to show and be beneficial would to be able to use that strategy where 2 are in each cell etc 

> the number of gyms that hold an EX Raid at one square is either 1 or 2 only.

showing the s2 cells for raids in combination with only showing ex raid potiential gyms would allow end users to easily ID all the gyms that should get raided at to maximize raid passes for that square :)

KartulUdus commented:
[x] raid and gym only switches 
[x] scanning doesn't overwrite park
[ ] ideas on how to query "Close enough" to geofence 
[ ] weird behaviour where raid egg image disappears if `Park Raids Only` activated

jorgitoh commented:
I have a problem here:

![map1osm](https://user-images.githubusercontent.com/87159/35253559-d414e3c6-ffc4-11e7-9020-984d150864ff.png)
![map2osm](https://user-images.githubusercontent.com/87159/35253561-d875fc20-ffc4-11e7-81e5-be3ca6f33c2e.png)

Both gyms are not considered (or are not inside) with ur code.

darkelement1987 commented:
You ran with -exg parameter?

jorgitoh commented:
@darkelement1987 yes

KartulUdus commented:
@jorgitoh 
I did a check on this area and got that 

Monumento a los Congresales de Tucumán - is not eligible for ex raid 
However, Fuente Deprecada & Glorieta are 

the monument, although in seemingly a park area, does not have any of it's s2 lvl 20 cell corners inside the fence 

sources: 
http://overpass-turbo.eu/
http://s2.sidewalklabs.com/regioncoverer/?center=-34.559197%2C-58.451587&zoom=20
![screen shot 2018-01-23 at 09 01 29](https://user-images.githubusercontent.com/23346150/35262207-582d0dda-001c-11e8-8bcf-9f968c9ff5c6.png)
![screen shot 2018-01-23 at 09 02 00](https://user-images.githubusercontent.com/23346150/35262226-68dbedf4-001c-11e8-9dee-17cc4d7ee040.png)
![screen shot 2018-01-23 at 09 02 03](https://user-images.githubusercontent.com/23346150/35262227-6b2378ac-001c-11e8-953d-d31feaa876c5.png)
![screen shot 2018-01-23 at 09 02 40](https://user-images.githubusercontent.com/23346150/35262229-6d1dd03a-001c-11e8-8513-10248a9ba7c8.png)

jorgitoh commented:
Thank you. So all the s2 lvl 20 cell corners must be inside of the fence?.

Ur code doesnt return "Fuente Deprecada" as an elegible gym.

KartulUdus commented:
could be your geofence or not latest pull or gym not in database when running -exg
at least 1 cell corner needs to be inside a park for it to be counted as a park 

I ran it locally a few minutes before those screenshots and Fuente Deprecada & Glorieta were eligible in that park

sebastienvercammen commented:
Posted on Discord:

> [6:16 PM] p_brent: it was discovered that all EX raid are in level 20 cells which have the cell centre inside the park
> [6:16 PM] p_brent: not just border is contained in cell

![Example](https://i.imgur.com/5bGVNdl.png)

Source: https://www.reddit.com/r/TheSilphRoad/comments/7t5vn1/new_discovery_for_nonsponsored_ex_raid_eligible/

sebastienvercammen commented:
#2464 is now merged, meaning auto-increment IDs in the tables will now stay the same, since we no longer rely on peewee's `REPLACE INTO`.

pogo-excalibur commented on package.json:
```diff
@@ -5,7 +5,7 @@
   "dependencies": {
     "babel-preset-env": "^1.6.0",
-    "eslint": "^3.1.1",
+    "eslint": "^3.19.0",
```
Would you mind outlining the resaon for this?

KartulUdus replied:
whoopsie, must of done a `npm too many things` reverted now

pogo-excalibur commented on pogom/models.py:
```diff
@@ -622,6 +623,10 @@ def get_gym(id):
         return result
 
+    @staticmethod
+    def is_gym_park(id, park):
```
I don't know if this is just the convention I'm used to, but I expect `is_*` methods to fetch a Boolean property. This seems to be setting a property, so I'd probably use something like `link_gym_to_park` or `set_gym_park`.

pogo-excalibur replied:
It may also be worth discussing whether we should store *which* park (feature id) the gym is in.

KartulUdus replied:
method renamed as recommended. 

unfortunately, not all parks return a 'name' for the way. would now know how to exactly tackle that

pogo-excalibur commented on pogom/osm.py:
```diff
@@ -0,0 +1,107 @@
+app = None
+log = logging.getLogger(__name__)
+db = init_database(app)
+earthCircumferenceMeters = 1000 * 40075.017
```
Why perform a multiplication here?
And this should be a constant: `EARTH_CIRCUMFERENCE_METERS`.

KartulUdus replied:
Capitalized now

pogo-excalibur commented on pogom/osm.py:
```diff
@@ -0,0 +1,107 @@
+                                                        len(ex_gyms.ways)))
+
+    for gym in gyms.items():
+        gympoint = {'lat': float(gym[1]['latitude']),
```
Not a big issue, but why use a dict here rather than two variables?
Consistency?

KartulUdus replied:
I believe the geofencing function `def is_point_in_polygon_custom`
expects the point to have a 'lat' & 'lon' key

pogo-excalibur replied:
Right. But `gympoint` doesn't get passed to `is_point_in_polygon_custom` - it only gets passed to `get_s2_cell_center`.

pogo-excalibur commented on pogom/osm.py:
```diff
@@ -0,0 +1,107 @@
+                break
+
+
+def get_s2_cell_center(lat, lng, radius, parent_level):
```
This method seems remarkably complex for the task, particularly as `radius` seems to always be `0`.
Could you outline why getting the `CellId` of the `LatLng` object and then getting the cell's center isn't sufficient?

pogo-excalibur commented on static/js/map.js:
```diff
@@ -1797,12 +1802,26 @@ function processGym(i, item) {
     }
 
+    if (Store.get('showParkGymsOnly')) {
+        if (item.park === false) {
```
Consider `if (!item.park) {`.

sebastienvercammen replied:
☝️

sebastienvercammen replied:
Still missing this ☝️ (read previous comment)

pogo-excalibur commented on static/js/map.js:
```diff
@@ -1204,7 +1208,8 @@ function updateGymMarker(item, marker) {
         })
         marker.setZIndex(google.maps.Marker.MAX_ZINDEX + 1)
-    } else if (item.raid && item.raid.end > Date.now() && Store.get('showRaids') && !Store.get('showActiveRaidsOnly') && raidLevel >= Store.get('showRaidMinLevel') && raidLevel <= Store.get('showRaidMaxLevel')) {
+    } else if (item.raid && item.raid.end > Date.now() && Store.get('showRaids') && !Store.get('showActiveRaidsOnly') &&
```
You'll need to use something like: `&& (!Store.get('showParkRaidsOnly') || item.park)`.
(untested code - so treat with caution)

And it'll also need to be added to the first `if` level at L1201.

pogo-excalibur replied:
And I'd suggest adding it to the *end* of the current chain.

pogo-excalibur replied:
Map has been 'fixed', but the issue with raid eggs not appearing will now be back.
I'd suggest trying the code snippet I posted above.

pogo-excalibur commented on templates/map.html:
```diff
@@ -152,6 +162,16 @@ <h3>Open Spot</h3>
                 </div>
               </div>
+              <div class="form-control switch-container" id="park-gyms-only-wrapper">
+                <h3>Gyms In Park</h3>
```
Perhaps this could be something more informative like 'Only Show Gyms In Parks'?

darkelement1987 replied:
"Park Gyms Only" ?

Markair commented:
Just attempted to merge this and hit an error (as below), running latest version of RM. Looks as though it needs a default value setting for bool field (false).

2018-01-29 23:05:21,290 [        MainThread][     runserver][   ERROR] Uncaught exception
Traceback (most recent call last):
  File "runserver.py", line 579, in <module>
    main()
  File "runserver.py", line 330, in main
    db = startup_db(app, args.clear_db)
  File "runserver.py", line 216, in startup_db
    verify_database_schema(db)
  File "C:\xampp\htdocs\RocketMap\pogom\models.py", line 2920, in verify_database_schema
    if not database_migrate(db, db_ver):
  File "C:\xampp\htdocs\RocketMap\pogom\models.py", line 3262, in database_migrate
    migrator.add_column('gym', 'park', BooleanField()))
  File "C:\Python27\lib\site-packages\playhouse\migrate.py", line 684, in migrate
    operation.run()
  File "C:\Python27\lib\site-packages\playhouse\migrate.py", line 144, in run
    getattr(self.migrator, self.method)(*self.args, **kwargs))
  File "C:\Python27\lib\site-packages\playhouse\migrate.py", line 152, in inner
    return fn(self, *args, **kwargs)
  File "C:\Python27\lib\site-packages\playhouse\migrate.py", line 222, in add_column
    raise ValueError('%s is not null but has no default' % column_name)
ValueError: park is not null but has no default

pogo-excalibur commented on static/js/map.js:
```diff
@@ -1208,8 +1208,8 @@ function updateGymMarker(item, marker) {
-    } else if (item.raid && item.raid.end > Date.now() && Store.get('showRaids') && !Store.get('showActiveRaidsOnly') &&
-        raidLevel >= Store.get('showRaidMinLevel') && !Store.get('showParkRaidsOnly') && raidLevel <= Store.get('showRaidMaxLevel')) {
+    } else if (item.raid && item.raid.end > Date.now() && Store.get('showRaids') && !Store.get('showActiveRaidsOnly') || !Store.get('showParkRaidsOnly' &&
+        raidLevel >= Store.get('showRaidMinLevel') && raidLevel <= Store.get('showRaidMaxLevel'))) {
```
What's happened here :) ?

pogo-excalibur commented on static/js/map.js:
```diff
@@ -1204,7 +1208,8 @@ function updateGymMarker(item, marker) {
         })
         marker.setZIndex(google.maps.Marker.MAX_ZINDEX + 1)
-    } else if (item.raid && item.raid.end > Date.now() && Store.get('showRaids') && !Store.get('showActiveRaidsOnly') && raidLevel >= Store.get('showRaidMinLevel') && raidLevel <= Store.get('showRaidMaxLevel')) {
+    } else if (item.raid && item.raid.end > Date.now() && Store.get('showRaids') && !Store.get('showActiveRaidsOnly') || !Store.get('showParkRaidsOnly' &&
```
What's happened here 😄 ?

KartulUdus replied:
it breaks 🙄

pogo-excalibur commented on static/js/map.js:
```diff
@@ -1204,7 +1208,8 @@ function updateGymMarker(item, marker) {
         })
         marker.setZIndex(google.maps.Marker.MAX_ZINDEX + 1)
-    } else if (item.raid && item.raid.end > Date.now() && Store.get('showRaids') && !Store.get('showActiveRaidsOnly') && raidLevel >= Store.get('showRaidMinLevel') && raidLevel <= Store.get('showRaidMaxLevel')) {
+    } else if (item.raid && item.raid.end > Date.now() && Store.get('showRaids') && !Store.get('showActiveRaidsOnly') &&
```
The map has been 'fixed', but the issue with raid eggs not appearing will now be back.

You'll need to use something like: `(!Store.get('showParkRaidsOnly') || item.park)`.
(untested code - so treat with caution)

And it'll also need to be added to the first `if` level at L1201.

KartulUdus replied:
when added to 1201, it never removed the !park gyms 

I believe this is resolved now with 
` && (!Store.get('showActiveRaidsOnly') || !Store.get('showParkRaidsOnly')) &&`

sebastienvercammen commented:
@KartulUdus Can you rebase?

fosJoddie commented on pogom/utils.py:
```diff
@@ -496,6 +496,10 @@ def get_args():
                         action='store_true', default=False)
+    parser.add_argument('-exg', '--ex-gyms',
+                        help=('Fetch OSM parks within geofence and ' +
+                              'mark gyms that are in sed areas'),
```
possible typo? I suggest changing to `Fetch OSM parks within geofence and flag gyms that are candidates for ex raids.`

KartulUdus replied:
done + also added to help= that `Only required once per area.`

KartulUdus replied:
Also added periods to comments and log messages. 

Holding fire on documentation until specified where to what.

sebastienvercammen commented on docs/extras/exclusive-raids.md:
```diff
@@ -0,0 +1,58 @@
+# Exclusive Raids
+
+EX-raids in Pokemon Go appear at gyms in parks, playgrounds and leisure areas.
```
> Pokémon

KartulUdus replied:
✅

sebastienvercammen commented on docs/extras/exclusive-raids.md:
```diff
@@ -0,0 +1,58 @@
+## How to check your area
+
+  - Scan your area normally until you have all gyms you wish to check.
+  - Run a single instance with a [geofence](geofences.md) and the `-exg` flag.  
```
Include the long form the first time, for clarity.

> `-exg` (`--ex-gyms`)

KartulUdus replied:
✅

sebastienvercammen commented on docs/extras/exclusive-raids.md:
```diff
@@ -0,0 +1,58 @@
+## Extra
+
+More specifically, a gym that could spawn an exclusive raid will have the center of its s2 level 20 cell inside a confirmed park area.
+You can manually check for s2 cells [here](http://s2.sidewalklabs.com/regioncoverer/).  
```
http://s2.sidewalklabs.com/regioncoverer/

![image](https://user-images.githubusercontent.com/4874269/36351109-c9d10364-14a4-11e8-8562-7a57fd54989c.png)

KartulUdus replied:
👌 
This was intended as an extra, should someone not trust the outcome of the check.

sebastienvercammen commented on pogom/models.py:
```diff
@@ -2112,7 +2118,15 @@ def parse_map(args, map_dict, scan_coords, scan_location, db_update_queue,
+                    park = Gym.select(Gym.park).where(
+                        Gym.gym_id == f.id).dicts()[0]['park']
+                except IndexError:
+                    park = False
```
Don't need to rely on the exception, you can just use the dict's `.get('key', default)`:

```
park = Gym.select(Gym.park).where(Gym.gym_id == f.id).dicts()[0].get('park', false)
```

KartulUdus replied:
✅

sebastienvercammen commented on pogom/models.py:
```diff
@@ -2112,7 +2118,15 @@ def parse_map(args, map_dict, scan_coords, scan_location, db_update_queue,
                 b64_gym_id = str(f.id)
                 gym_display = f.gym_display
                 raid_info = f.raid_info
+                # try to get recorded value for Gyms.park, default to false.
```
Capitalize

> Try

KartulUdus replied:
✅

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,97 @@
+from utils import get_args
+from s2sphere import LatLng, CellId
+
+args = get_args
```
* `args` is unused.
* Should have been `args = get_args()` if you'd have used it.

KartulUdus replied:
😊 
...
✅

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,97 @@
+app = None
+log = logging.getLogger(__name__)
+db = init_database(app)
+EARTH_CIRCUMFERENCE_METERS = 400750170
```
What are `app` and `db` for? Unused.

KartulUdus replied:
🙅‍♂️ 
it's needed for gyms = Gym.get_gyms
without init_database, it can't query where existing gyms are.
see Friscos review above

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,97 @@
+def exgyms(geofence):
+    # Parse geofence file.
+    log.info('Finding border points from geofence.')
+    f = json.loads(json.dumps(Geofences.parse_geofences_file(geofence, '')))
```
* Use a more descriptive name than `f` (e.g. `geofences`).
* Doing both `json.loads(json.dump())` will have the original as result, so why do any at all?

KartulUdus replied:
✅

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,97 @@
+            if Geofences.is_point_in_polygon_custom(s2_center, data):
+                # Try to get Gym name, but default to id if missing.
+                try:
+                    gymname = Gym.get_gym(gym[0])['name'].encode('utf8')
```
No need to re-query the database.

`Gym.get_gyms(...)` already returns all gym details, including gym name.

KartulUdus replied:
✅

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,97 @@
+    return result
+
+
+def exgyms(geofence):
```
* Function name `exgyms` isn't descriptive enough.
* I would separate the function into several separate steps so we don't have one big blob that's supposed to do everything, e.g.:
```
# Assumes .get_first_geofence is added to Geofences.
def update_ex_gyms(geofence):
    fence = Geofence.get_first_geofence()
    ...
    ex_gyms = ex_query(south, west, north, east)
    gyms = Gym.get_gyms(south, west, north, east)
    ...
    confirmed_ex_gyms = filter(__gym_is_ex_gym, gyms)
    # Assumes .set_gyms_in_park is added to Gym.
    Gym.set_gyms_in_park(confirmed_ex_gyms, True)


def __gym_is_ex_gym(gym, ex_gyms):
    ex_gyms = ex_gyms.ways

    for way in ex_gyms.ways:
        ... # Copy code here that's currently in exgyms.
        return True

    return False
```

KartulUdus replied:
✅

sebastienvercammen commented on pogom/models.py:
```diff
@@ -626,6 +628,12 @@ def get_gym(id):
+    def set_gyms_in_park(gyms, park):
+        for gym in gyms:
+            Gym.update(park=park).where(
+                Gym.gym_id == str(gym['gym_id'])).execute()
```
Replace with:

```
# Instead of doing this, you can also only pass the gym_ids as an argument rather than gyms. Up to you.
gym_ids = [gym['gym_id'] for gym in gyms]
# Execute a single query on a list rather than running a query for each item in the list.
Gym.update(park=park).where(Gym.gym_id << gym_ids).execute()
```

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,93 @@
+    [timeout:620]
+    [bbox:{},{},{},{}];
+    (
+    //Tags that are confirmed to classify gyms as 'parks' for EX Raids
```
Write comments as:

> // Capitalized sentence, with punctuation.

*(space and dot)*

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,93 @@
+    log.info('Checking {} gyms against {} parks.'.format(len(gyms),
+                                                         len(ex_gyms.ways)))
+    confirmed_ex_gyms = filter(lambda gym: __gym_is_ex_gym(gym, ex_gyms),
+                               gyms.values())
```
Add an empty line and a comment above the `filter` call for clarity:

```
# Retrieve list of confirmed EX gyms to update the DB.
confirmed_ex_gyms = ...
```

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,93 @@
+    log.info('Finding border points from geofence.')
+    geofence = Geofences.parse_geofences_file(geofence, '')
+    fence = geofence[0]['polygon']
+    # Figure out borders for bounding box.
```
Add an empty line before this comment.

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,93 @@
+    west = min(fence, key=lambda ev: ev['lon'])['lon']
+    north = max(fence, key=lambda ev: ev['lat'])['lat']
+    east = max(fence, key=lambda ev: ev['lon'])['lon']
+    log.info('Finding parks within zone.')
```
Add an empty line before this log.

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,93 @@
+    s2_center = get_s2_cell_center(gympoint[0], gympoint[1], 20)
+    for way in ex_gyms.ways:
+        data = []
+        for node in way.nodes:
```
Empty line before this for loop (separate blocks).

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,93 @@
+        for node in way.nodes:
+            data.append({'lat': float(node.lat),
+                         'lon': float(node.lon)})
+        if Geofences.is_point_in_polygon_custom(s2_center, data):
```
Empty line before this `if`.

KartulUdus commented:
I think that's the all of it now @sebastienvercammen 

- spaces punctuation line breaks and stuff
- update by array after -exg check 
- found and fixed a nasty where console would error when finding a new gym 
- updated query and doc to include farmyard 
- added park as `true|false` in `gym` webhook ( I want to `!raid snorlax park d2000` )

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,102 @@
+
+app = None
+log = logging.getLogger(__name__)
+db = init_database(app)
```
Add a comment above this to clarify that you're doing it for side effects (= the effect of the function that isn't a return value).

KartulUdus replied:
done

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,102 @@
+    for way in ex_gyms.ways:
+        data = []
+        for node in way.nodes:
+            data.append({'lat': float(node.lat),
```
The block `for way in ex_gyms.ways:` will be done every time `__gym_is_ex_gym` is called, and it'll always be on the same data, so it'll unnecessarily be repeated.

Build the `way.nodes` data list outside of the function instead, and pass it as `__gym_is_ex_gym(gym, way_nodes)` instead so it's only done once.

KartulUdus replied:
no.

I moved that loop you don't like outside __gym_is_ex_gym alas, then I iterated through the list twice. 

that would of not worked regardless as an overpy impediment there gave me an error later:
``
    ex_gyms.ways[index]['polygon'] = data
TypeError: 'Way' object does not support item assignment
```

this was by moving the thing *Back* to update_ex_gyms where it resided before.

```
# Generate polygons that is_point_in_polygon_custom accepts for each park.
for index, way in enumerate(ex_gyms.ways):
    data = [{'lat': float(e.lat), 'lon': float(e.lon)} for e in way.nodes]
    ex_gyms.ways[index]['polygon'] = data
```
Please see an [example of a result](https://hastebin.com/ihimihafut.json) from overpy 
the object returned from overpy is a scattered mess of nodes and to know which nodes make up a park, you need to iterate over the "way" first 

I did make the for loop prettier inside __gym_is_ex_gym and ran a test against an area

pogo-excalibur commented on docs/extras/exclusive-raids.md:
```diff
@@ -0,0 +1,62 @@
+# Exclusive Raids
```
Niantic seem to have switched from using the term 'exclusive' to *only* using the term 'EX'.
I think it would be reasonable for us to follow this.

pogo-excalibur commented on pogom/osm.py:
```diff
@@ -0,0 +1,100 @@
+
+        if Geofences.is_point_in_polygon_custom(s2_center, data):
+            gymname = gym['name'] or gym['gym_id']
+            log.info('{} is eligible for EX raid.'.format(
```
'is eligible for EX raid**s**' (add an 's' on the end)

pogo-excalibur replied:
Or perhaps better:
```
Found {count} EX-elegible gyms:
> Gym 1
> Gym 2
> ...
```

pogo-excalibur commented on pogom/utils.py:
```diff
@@ -496,6 +496,11 @@ def get_args():
                         action='store_true', default=False)
+    parser.add_argument('-exg', '--ex-gyms',
+                        help=('Fetch OSM parks within geofence and flag ' +
+                              'gyms that are candidates for ex raids. ' +
```
'gyms that are candidates for ex' -> 'gyms that are candidates for EX'

pogo-excalibur commented on pogom/utils.py:
```diff
@@ -1309,7 +1314,6 @@ def get_debug_dump_link():
     result = '''#######################
 ### RocketMap debug ###
 #######################
-
```
Is this intentional?

pogo-excalibur commented:
Question for discussion:
Should we remove gyms that are listed as being in parks, but which have the attribute `enabled: false` ?

In my experience these gyms do not spawn raids, and so cannot host an EX raid.

pogo-excalibur commented on docs/extras/exclusive-raids.md:
```diff
@@ -0,0 +1,62 @@
+
+RocketMap includes a built-in tool to check if existing gyms fall inside these areas.
+
+## How to check your area
```
Add a note in here that only the first region in the geofence file will get checked.

pogo-excalibur commented:
Issue while testing with no ex-eligible gyms nearby: `ProgrammingError: (1064, u"You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '))' at line 1")`

This is caused by:
```
@staticmethod
def set_gyms_in_park(gyms, park):
    gym_ids = [gym['gym_id'] for gym in gyms]
    Gym.update(park=park).where(Gym.gym_id << gym_ids).execute()
```

With no ex-eligible gyms you get: `gyms = []`.

So the SQL query syntax:
```
UPDATE 'gym' SET 'park' = %s WHERE ('gym'.'gym_id' IN ())
```

becomes invalid because of the `IN ()` at the end.

pogo-excalibur commented:
The bug I found above is still present.

You've handled the case where there are *no gyms at all*, but that's not waht was causing the issue.

The bug occurs when there *are* gyms in the geofence, but *none* of them is EX-eligible.

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,109 @@
+from models import Gym, init_database
+from s2sphere import LatLng, CellId
+
+# Specify app for a database connection and log to show module name in logs
```
Dot.

KartulUdus replied:
.

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,109 @@
+
+def ex_query(s, w, n, e):
+
+    # Query Overpass for known gym areas.
```
No empty line above this comment.

KartulUdus replied:
✅

sebastienvercammen commented on pogom/models.py:
```diff
@@ -2112,7 +2119,15 @@ def parse_map(args, map_dict, scan_coords, scan_location, db_update_queue,
+                gym_by_id = Gym.select(Gym.park).where(
+                    Gym.gym_id == f.id).dicts()
+                if gym_by_id:
+                    park = gym_by_id[0]['park']
```
These 5 lines could be simplified as a new method on `Gym`, e.g. `Gym.is_in_park()` which returns `True` or `False`. Can return `False` if the `Gym` isn't in the DB.

For proper abstractions, you would implement a method that accepts a list of gyms and does a single `IN ()` query, then use that method to query for single gyms.

```
get_all(x):
    return SELECT * ... IN (x)

get_one(x):
    return get_all(x)

or:

get_one(x):
    return SELECT * ... WHERE x = x
```
(Depends on whether `IN` versus `=` has any performance effect on single items.)

Up to you.

KartulUdus replied:
it's ok. We're over-engineering a simple task that the end-human would only run once (per -cd) for a minute or so

sebastienvercammen replied:
@KartulUdus Code style and proper abstractions aren't overengineering.

In this case it's OK to leave as is since no one else needs it (yet), but the moment we need the same code anywhere else, we'll have to implement it anyway.

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,109 @@
+    log.info('Finding parks within zone.')
+    ex_gyms = ex_query(south, west, north, east)
+    gyms = Gym.get_gyms(south, west, north, east)
+    if (gyms):
```
Don't surround `gyms` with parentheses.

KartulUdus replied:
new habits develop swift

sebastienvercammen commented on pogom/osm.py:
```diff
@@ -0,0 +1,109 @@
+            Gym.set_gyms_in_park(confirmed_ex_gyms, True)
+        else:
+            log.info('No EX-eligible gyms found.')
+    else:
```
To solve nesting

```
if x:
    if y:
        a()
    else:
        b()
else:
    c()
```

use this code style:

```
# No gyms at all.
if not gyms:
    a()

confirmed_ex_gyms = ...

# No EX-eligible gyms.
if not confirmed_ex_gyms:
    b()

# At this point, we have a set of EX-eligible gyms and we can continue.
```

sebastienvercammen commented on runserver.py:
```diff
@@ -330,6 +331,15 @@ def main():
+        if args.geofence_file == '':
+            log.critical('A geofence is required to find EX-gyms.')
+            sys.exit(1)
+        else:
```
No need for this `else`: the first branch of the conditional exits the process, so anything after that line of code is assured to have a geofence file and `-exg` enabled.

KartulUdus replied:
Thanks, makes sense 
✅

sebastienvercammen commented on runserver.py:
```diff
@@ -330,6 +331,15 @@ def main():
     args.root_path = os.path.dirname(os.path.abspath(__file__))
 
+    if args.ex_gyms:
+        if args.geofence_file == '':
```
Replace this line with:

```
# Geofence is required.
if not args.geofence_file:
```

KartulUdus replied:
I thought I tried that first to no avail 🤔 
anyway: 
✅

sebastienvercammen commented on static/js/map.js:
```diff
@@ -1204,7 +1208,8 @@ function updateGymMarker(item, marker) {
         marker.setZIndex(google.maps.Marker.MAX_ZINDEX + 1)
-    } else if (item.raid && item.raid.end > Date.now() && Store.get('showRaids') && !Store.get('showActiveRaidsOnly') && raidLevel >= Store.get('showRaidMinLevel') && raidLevel <= Store.get('showRaidMaxLevel')) {
+    } else if (item.raid && item.raid.end > Date.now() && Store.get('showRaids') && (!Store.get('showActiveRaidsOnly') || !Store.get('showParkRaidsOnly')) &&
+        raidLevel >= Store.get('showRaidMinLevel') && raidLevel <= Store.get('showRaidMaxLevel')) {
```
This `if` is way too long. Clarify:
```
const hasActiveRaid = item.raid && item.raid.end > Date.now()
const raidLevelVisible = raidLevel >= Store.get('showRaidMinLevel') && raidLevel <= Store.get('showRaidMaxLevel')
const showRaidSetting = Store.get('showRaids') && (!Store.get('showActiveRaidsOnly') || !Store.get('showParkRaidsOnly'))

if (hasActiveRaid && raidLevelVisible && showRaidSetting) {
    ...
}
```

KartulUdus replied:
did the things. ✅ 

Thought maybe to change `let raidLevel = getRaidLevel(item.raid)` to also be a const, alas i'm not sure if it get's reassigned somewhere of if const would break it so I left it as `let`.
let, var, const is used on map.js depending on moon phases and wind direction so better not touch.
